src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/AbstractQueryAndScanMapper.java

```diff
@@ -0,0 +1,292 @@
+        .put(ComparisonOperator.LT, LegacyKeyConditionType.LT)
+        .put(ComparisonOperator.LE, LegacyKeyConditionType.LE)
+        .put(ComparisonOperator.BETWEEN, LegacyKeyConditionType.BETWEEN)
+        .build();
+
+    /*
+     * Converts QueryRequest objects containing keyConditions to keyConditionExpression and ScanRequest objects
+     * containing scanFilters.  According to the DynamoDB docs, QueryRequest keyConditions and ScanRequest scanFilter
+     * fields are considered 'legacy parameters'.  However, we support them by converting them to keyConditionExpression
+     * and filterExpression fields respectively because they are used by the DynamoDB document API
+     * (https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/dynamodbv2/document/DynamoDB.html).
+     */
+    private void convertLegacyExpression(RequestWrapper request, Map<String, Condition> legacyConditions) {
+        if (legacyConditions != null && !legacyConditions.isEmpty()) {
+            List<String> keyConditionExpressionParts = new ArrayList<>();
+            legacyConditions.forEach((key, condition) -> {
+                final ComparisonOperator comparisonOperator = ComparisonOperator
+                    .valueOf(condition.getComparisonOperator());
+                LegacyKeyConditionType type = KEY_CONDITION_PER_OPERATOR.get(comparisonOperator);
+                checkNotNull(type,"unsupported comparison operator " + condition.getComparisonOperator()
+                    + " in condition=" + condition);
+                checkArgument(condition.getAttributeValueList().size() == type.numValues,
+                    "keyCondition with wrong number of attribute values (" + condition.getAttributeValueList().size()
+                        + ") encountered in condition=" + condition);
+
+                String fieldPlaceholder = getNextFieldPlaceholder(request);
+                request.putExpressionAttributeName(fieldPlaceholder, key);
+
+                List<String> valuePlaceholders = new ArrayList<>(condition.getAttributeValueList().size());
+                condition.getAttributeValueList().forEach(value -> {
+                    String valuePlaceholder = getNextValuePlaceholder(request);
+                    request.putExpressionAttributeValue(valuePlaceholder, value);
+                    valuePlaceholders.add(valuePlaceholder);
+                });
+
+                String expression = type.toExpression.apply(fieldPlaceholder, valuePlaceholders);
+                keyConditionExpressionParts.add(expression);
+            });
+            request.setExpression(Joiner.on(" AND ").join(keyConditionExpressionParts));
+        }
+    }
+
+    interface QueryOrScanRequestWrapper extends RequestWrapper {
+
+        Map<String, AttributeValue> getExclusiveStartKey();
+
+        void setExclusiveStartKey(Map<String, AttributeValue> exclusiveStartKey);
+
+        String getIndexName();
+
+        void setIndexName(String indexName);
+    }
+
```

Comment:
I also did some refactoring of RequestWrapper, so that each wrapper is used to get/set at most 1 expression, instead of possibly 2 expressions (primaryExpression and filterExpression). This seems less confusing to me.